app/grandchallenge/algorithms/tasks.py

```diff
@@ -163,105 +159,137 @@ def create_algorithm_jobs(
+    archives = {ai.archive for ai in archive_items}
+    if len(archives) != 1:
+        raise RuntimeError(
+            "The archive items should all belong to the same archive."
         )
-        .annotate(
-            inputs_match_count=Count(
-                "inputs",
-                filter=Q(
-                    inputs__in={civ for civ_set in civ_sets for civ in civ_set}
-                ),
-            )
-        )
-        .filter(inputs_match_count=len(input_interfaces), creator=None)
-        .prefetch_related("inputs")
-    }
-
-    valid_job_inputs = []
-
-    for civ_set in civ_sets:
-        # Check interfaces are complete
-        civ_interfaces = {civ.interface for civ in civ_set}
-        if input_interfaces.issubset(civ_interfaces):
-            # If the algorithm works with a subset of the interfaces
-            # present in the set then only feed these through to the algorithm
-            valid_input = {
-                civ for civ in civ_set if civ.interface in input_interfaces
-            }
-        else:
-            continue
+    (archive,) = archives
 
-        # Check job has not been run
-        if frozenset(valid_input) in existing_jobs:
-            continue
+    algorithm_interfaces = (
+        algorithm_image.algorithm.interfaces.prefetch_related("inputs").all()
+    )
 
-        valid_job_inputs.append(valid_input)
+    # First, sort archive items by algorithm interface:
+    # An archive item is only valid for an interface if it has values
+    # for all inputs of the interface
+    valid_job_inputs = get_archive_items_for_interfaces(
+        algorithm_interfaces=algorithm_interfaces, archive_items=archive_items
+    )
+
+    # Next, get all system jobs that have been run for the provided archive items
+    # with the same model and image
+    if algorithm_model:
+        extra_filter = {"algorithm_model": algorithm_model}
+    else:
+        extra_filter = {"algorithm_model__isnull": True}
+
+    relevant_jobs = Job.objects.filter(
+        algorithm_image=algorithm_image,
+        algorithm_interface__in=valid_job_inputs.keys(),
+        creator=None,
+        inputs__archive_items__archive=archive,
+        **extra_filter,
+    )
+    # and group those by interface
+    existing_jobs_for_interfaces = (
+        get_valid_jobs_for_interfaces_and_archive_items(
+            algorithm_interfaces=algorithm_interfaces,
+            jobs=relevant_jobs,
+            valid_archive_items_per_interface=valid_job_inputs,
+        )
+    )
+
+    # Finally, exclude archive items for which there already is a job
+    filtered_valid_job_inputs = {}
+    for interface, archive_items in valid_job_inputs.items():
+        items_with_job = [
+            ai.pk
+            for ai in archive_items
+            if frozenset(ai.values.all())
+            in [
+                frozenset(j.inputs.all())
+                for j in existing_jobs_for_interfaces[interface]
+            ]
```

Comment:
It may or may not matter here but doing set lookups is quicker as the items are hashed, so python does not need to iterate over the items and perform a comparison each time.

```suggestion
            in {
                frozenset(j.inputs.all())
                for j in existing_jobs_for_interfaces[interface]
            }
```

Also double check that this isn't being repeatedly calculated. It may be better to extract this to a variable which is declared before `items_with_job`.

Reply:
I double checked and can confirm that this is not repeatedly calculated. `existing_jobs_for_interfaces` contains an already evaluated list of jobs per interface.

Reply:
Actually, moving this out of the loop helps with readability and I also realized that I can directly filter for items withOUT a job and hence can avoid the later list comprehension all together.